presto-main/src/main/java/io/prestosql/sql/planner/optimizations/QueryCardinalityUtil.java

```diff
@@ -106,10 +106,28 @@ public CardinalityExtractorPlanVisitor(Lookup lookup)
         @Override
         public Range<Long> visitAggregation(AggregationNode node, Void context)
         {
-            if (node.hasEmptyGroupingSet()) {
+            if (node.hasEmptyGroupingSet() && node.getGroupingSetCount() == 1) {
+                // only single default aggregation which will produce exactly single row
                 return Range.singleton(1L);
             }
-            return Range.atLeast(0L);
+
+            Range<Long> sourceCardinalityRange = node.getSource().accept(this, null);
+
+            if (node.hasDefaultOutput() || sourceCardinalityRange.lowerEndpoint() > 0) {
+                if (sourceCardinalityRange.hasUpperBound()) {
+                    return Range.closed(1L, sourceCardinalityRange.upperEndpoint());
```

Comment:
`sourceCardinalityRange` can be `[0, 0]`, this will fail then.

Reply:
good point